File: src/localcoder_cli/codebase.py
from __future__ import annotations

import subprocess
from pathlib import Path


def _run(cmd: list[str], cwd: Path) -> str:
    try:
        out = subprocess.check_output(cmd, cwd=cwd, text=True, stderr=subprocess.STDOUT)
        return out.strip()
    except Exception:
        return ""


def summarize_repo(root: Path) -> str:
    files = _run(["git", "ls-files"], root).splitlines()
    if not files:
        return "No git-tracked files found."

    top = files[:40]
    languages = {}
    for f in files:
        suffix = Path(f).suffix.lower() or "[no_ext]"
        languages[suffix] = languages.get(suffix, 0) + 1

    lang_summary = ", ".join(
        f"{k}:{v}" for k, v in sorted(languages.items(), key=lambda x: x[1], reverse=True)[:10]
    )

    return (
        f"Repository root: {root}\n"
        f"Tracked files: {len(files)}\n"
        f"Top extensions: {lang_summary}\n"
        "Sample files:\n"
        + "\n".join(f"- {x}" for x in top)
    )

File: src/codebase.js
import { execSync } from "node:child_process";
import path from "node:path";

function safeExec(command, cwd) {
  try {
    return execSync(command, { cwd, stdio: ["ignore", "pipe", "ignore"], encoding: "utf8" }).trim();
  } catch {
    return "";
  }
}

export function summarizeRepo(rootDir) {
  const filesRaw = safeExec("git ls-files", rootDir);
  const files = filesRaw ? filesRaw.split("\n").filter(Boolean) : [];
  if (files.length === 0) {
    return "No git-tracked files found.";
  }

  const extensionCounts = {};
  for (const file of files) {
    const ext = path.extname(file).toLowerCase() || "[no_ext]";
    extensionCounts[ext] = (extensionCounts[ext] || 0) + 1;
  }

  const topExtensions = Object.entries(extensionCounts)
    .sort((a, b) => b[1] - a[1])
    .slice(0, 10)
    .map(([ext, count]) => `${ext}:${count}`)
    .join(", ");

  const sampleFiles = files.slice(0, 40).map((entry) => `- ${entry}`).join("\n");
  return `Repository root: ${rootDir}
Tracked files: ${files.length}
Top extensions: ${topExtensions}
Sample files:
${sampleFiles}`;
}
